### PR TITLE
Docker development guide

### PR DESCRIPTION
**Pull request by SteveMacenski**

### CODE REVIEW

**Review by Ryanf55 — COMMENTED** [review]
Super awesome to see something like this written up. Truly zero to hero.

I do think this is a great intro, however I would like to consider a "Production concerns" section that mentions:
* Security
* Making use of multistage builds
* Leaking source code to deployment
* Managing cache and avoiding busting it
* Levarging mounts to speed things up
* Usage of rosdep to only install runtime dependencies
* Building in Debug vs Release for production deploy and production debug
* Exporting test results from `colcon` back to CI with Junit

For now, this could go in, and perhaps we can just put a list of "Things to think about in production"

File: tutorials/docs/docker_dev.rst [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+  RUN apt-get update \
+      && apt-get install -y \
+    ros-dev-tools \
+    wget
+    
+  WORKDIR /root/nav2_ws 
+  RUN mkdir -p ~/nav2_ws/src
+  RUN git clone https://github.com/ros-planning/navigation2.git --branch main ./src/navigation2
+  RUN rosdep init
+  RUN apt update && apt upgrade -y \
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply tutorials/docs/docker_dev.rst:434]: What warnings?

File: tutorials/docs/docker_dev.rst [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+Its useful to note at this point that the ``--privileged`` flag is a real hammer. If you want to avoid running this, you can find all the individual areas you need to enable for visualization to work.
+Also note that ``--privileged`` also makes it easier to run hardware interfaces like joysticks and sensors by enabling inputs from the host operating system that are processing those inputs.
+If in production, you cannot use a hammer, you may need to dig into your system a bit to allow through only the interfaces required for your hardware.
+
+As for potential steps forward: 
+- Setup a config file to hide all those docker run arguments for development
+- Setup a bash script to enable several different configurations of docker run and execute the run itself
+- Learn more about Docker's options and features such as compose, pushing your own containers to DockerHub, and version controlling images
+- Limit and regulate host resource utilization
+- Configure computer to avoid use of ``sudo`` for each docker CLI command.
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply tutorials/docs/docker_dev.rst:405]: Sure, what would you want here?

File: tutorials/docs/docker_dev.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply tutorials/docs/docker_dev.rst:434]: These show up in a non-interactive session, such as when building in CLI.

File: tutorials/docs/docker_dev.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply tutorials/docs/docker_dev.rst:405]: * https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user
* https://docs.docker.com/compose/  - Make sure to call it Docker Compose V2 btw.

File: tutorials/docs/docker_dev.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply tutorials/docs/docker_dev.rst:434]: https://askubuntu.com/questions/990823/apt-gives-unstable-cli-interface-warning

File: tutorials/docs/docker_dev.rst [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+  FROM ros:${ROS_DISTRO}-ros-core
+
+  RUN apt-get update \
+      && apt-get install -y \
+    ros-dev-tools \
+    wget
+    
+  WORKDIR /root/nav2_ws 
+  RUN mkdir -p ~/nav2_ws/src
+  RUN git clone https://github.com/ros-planning/navigation2.git --branch main ./src/navigation2
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply tutorials/docs/docker_dev.rst:432]: This isn't private though?

File: tutorials/docs/docker_dev.rst [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+
+  # For Rolling or want to build from source a particular branch / fork
+  WORKDIR /root/nav2_ws 
+  RUN mkdir -p ~/nav2_ws/src
+  RUN git clone https://github.com/ros-planning/navigation2.git --branch main ./src/navigation2
+  RUN rosdep init
+  RUN apt update && apt upgrade -y \
+      && rosdep update \
+      && rosdep install -y --ignore-src --from-paths src -r
+  RUN colcon build --symlink-install
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply tutorials/docs/docker_dev.rst:462]: I required `sudo` to run locally unless I added docker `usermod -aG docker $USER`. My understanding is that it still has the same `sudo` permissions, just I didn't need to type sudo. So what's the functional  difference?

File: tutorials/docs/docker_dev.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply tutorials/docs/docker_dev.rst:432]: No, NAV2 is a public repo. My company prohibits https cloning for example, and you have to have special flags to docker to access the SSH agent.

File: tutorials/docs/docker_dev.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply tutorials/docs/docker_dev.rst]: I'll put a section up for handling root permissions. 
https://dockerlabs.collabnix.com/security/Running-Containers-as-ROOT.html#:~:text=One%20of%20the%20best%20practices,root%20user%20access%20on%20host.

https://docs.docker.com/develop/develop-images/instructions/#user

**Pull request merged**